File: skills/supermap-jira/scripts/search_jira.py
#!/usr/bin/env python3
"""
Supermap Jira Search Script

Searches the Supermap Jira system and returns results as a markdown table.
Uses the SUPERMAP_JIRA_TOKEN environment variable for authentication.
"""

import os
import sys
import json
import urllib.request
import urllib.error
import urllib.parse


JIRA_BASE_URL = "https://jira.supermap.work"
API_ENDPOINT = f"{JIRA_BASE_URL}/rest/quicksearch/1.0/productsearch/search"


def get_token():
    """Get the Jira token from environment variable."""
    token = os.environ.get("SUPERMAP_JIRA_TOKEN")
    if not token:
        print("Error: SUPERMAP_JIRA_TOKEN environment variable is not set.", file=sys.stderr)
        print("Please set it with: export SUPERMAP_JIRA_TOKEN='your-token-here'", file=sys.stderr)
        sys.exit(1)
    return token


def search_jira(query, token):
    """
    Search Jira for the given query.

    Args:
        query: Search keyword(s)
        token: Bearer token for authentication

    Returns:
        Parsed JSON response as a dictionary
    """
    encoded_query = urllib.parse.quote(query)
    url = f"{API_ENDPOINT}?q={encoded_query}"

    request = urllib.request.Request(url)
    request.add_header("Authorization", f"Bearer {token}")
    request.add_header("Accept", "application/json")

    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            data = response.read().decode("utf-8")
            return json.loads(data)
    except urllib.error.HTTPError as e:
        if e.code == 401:
            print("Error: Authentication failed. Please check your SUPERMAP_JIRA_TOKEN.", file=sys.stderr)
        elif e.code == 403:
            print("Error: Access forbidden. You may not have permission to search.", file=sys.stderr)
        else:
            print(f"Error: HTTP {e.code} - {e.reason}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Error: Network error - {e.reason}", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError:
        print("Error: Failed to parse response from Jira.", file=sys.stderr)
        sys.exit(1)


def format_as_markdown_table(items):
    """
    Format the search results as a markdown table.

    Args:
        items: List of issue items with 'title', 'subtitle', and 'url' fields

    Returns:
        Markdown formatted string
    """
    if not items:
        return "No issues found."

    # Table header
    lines = ["| 标题 | 链接 |", "| --- | --- |"]

    # Table rows
    for item in items:
        title = item.get("title", "N/A")
        url = item.get("url", "")

        if url:
            lines.append(f"| {title} | {url} |")
        else:
            lines.append(f"| {title} | N/A |")

    return "\n".join(lines)


def extract_issues(response):
    """
    Extract only the 'quick-search-issues' results from the response.

    Args:
        response: Parsed JSON response

    Returns:
        List of issue items
    """
    if not isinstance(response, list):
        return []

    for section in response:
        if isinstance(section, dict) and section.get("id") == "quick-search-issues":
            return section.get("items", [])

    return []


def print_help():
    """Print help message."""
    help_text = """Supermap Jira Search

Usage: python search_jira.py <search-query>

Description:
    Search the Supermap Jira system and return matching issues as a markdown table.

Environment Variables:
    SUPERMAP_JIRA_TOKEN  - Required. Your Jira API token for authentication.

Examples:
    python search_jira.py iServer
    python search_jira.py "bug fix"
    python search_jira.py --help

Output:
    Results are displayed as a markdown table with issue titles and links.
"""
    print(help_text)


def main():
    """Main entry point."""
    args = sys.argv[1:]

    if not args or args[0] in ["--help", "-h"]:
        print_help()
        sys.exit(0)

    query = " ".join(args)
    token = get_token()

    response = search_jira(query, token)
    issues = extract_issues(response)

    output = format_as_markdown_table(issues)
    print(output)


if __name__ == "__main__":
    main()

File: skills/supermap-jira/scripts/search_jira.js
#!/usr/bin/env node
/**
 * Supermap Jira Search Script (Node.js Version)
 *
 * Searches the Supermap Jira system using JQL API and returns results as a markdown table.
 * Uses the SUPERMAP_JIRA_TOKEN environment variable for authentication.
 */

const https = require('https');

const JIRA_BASE_URL = 'jira.supermap.work';
const API_ENDPOINT = '/rest/api/2/search';

function getToken() {
    const token = process.env.SUPERMAP_JIRA_TOKEN;
    if (!token) {
        console.error('Error: SUPERMAP_JIRA_TOKEN environment variable is not set.');
        console.error('Please set it with:');
        console.error('  Linux/macOS: export SUPERMAP_JIRA_TOKEN=\'your-token-here\'');
        console.error('  Windows (cmd): set SUPERMAP_JIRA_TOKEN=your-token-here');
        console.error('  Windows (PowerShell): $env:SUPERMAP_JIRA_TOKEN=\'your-token-here\'');
        process.exit(1);
    }
    return token;
}

function makeRequest(url, token) {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: JIRA_BASE_URL,
            path: url,
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Accept': 'application/json',
                'User-Agent': 'Supermap-Jira-Search/1.0'
            },
            timeout: 30000,
            rejectUnauthorized: false
        };

        const req = https.request(options, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                if (res.statusCode >= 200 && res.statusCode < 300) {
                    try {
                        const jsonData = JSON.parse(data);
                        resolve(jsonData);
                    } catch (e) {
                        reject(new Error(`Failed to parse response: ${e.message}`));
                    }
                } else if (res.statusCode === 401) {
                    reject(new Error('Authentication failed. Please check your SUPERMAP_JIRA_TOKEN.'));
                } else if (res.statusCode === 403) {
                    reject(new Error('Access forbidden. You may not have permission to search.'));
                } else {
                    reject(new Error(`HTTP ${res.statusCode}`));
                }
            });
        });

        req.on('error', (err) => {
            reject(new Error(`Network error: ${err.message}`));
        });

        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });

        req.end();
    });
}

function buildJql(query, useRawJql) {
    if (useRawJql) {
        return query;
    }
    // Free-text search: escape double quotes in the query
    const escaped = query.replace(/"/g, '\\"');
    return `(summary ~ "${escaped}" OR description ~ "${escaped}") ORDER BY created DESC`;
}

async function searchJira(jql, token) {
    const encodedJql = encodeURIComponent(jql);
    const url = `${API_ENDPOINT}?jql=${encodedJql}&maxResults=30&fields=summary,status,priority,fixVersions,created`;

    try {
        const response = await makeRequest(url, token);
        return response;
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

function formatAsMarkdownTable(response) {
    const issues = response.issues || [];

    if (!issues || issues.length === 0) {
        return `No issues found. (total: ${response.total || 0})`;
    }

    const lines = ['| Key | 状态 | 优先级 | 版本 | 标题 |', '| --- | --- | --- | --- | --- |'];

    for (const issue of issues) {
        const key = issue.key || 'N/A';
        const fields = issue.fields || {};
        const summary = fields.summary || 'N/A';
        const status = fields.status ? fields.status.name : 'N/A';
        const priority = fields.priority ? fields.priority.name : 'N/A';
        const versions = (fields.fixVersions || []).map(v => v.name).join(', ') || '-';

        lines.push(`| ${key} | ${status} | ${priority} | ${versions} | ${summary} |`);
    }

    if (response.total > issues.length) {
        lines.push('');
        lines.push(`> 共 ${response.total} 条结果，显示前 ${issues.length} 条。`);
    }

    return lines.join('\n');
}

function printHelp() {
    console.log(`Supermap Jira Search (Node.js Version) - JQL API

Usage: node search_jira.js [--jql] <search-query>

Description:
    Search the Supermap Jira system using JQL and return matching issues.

    By default, the search query is treated as free-text, searching issue
    summary and description fields. Use --jql flag to pass raw JQL.

Environment Variables:
    SUPERMAP_JIRA_TOKEN  - Required. Your Jira API token for authentication.

Examples:
    node search_jira.js "范围查询"
    node search_jira.js --jql "(summary ~ \"范围查询\" OR summary ~ \"BOUNDS\") AND project = ISVJ ORDER BY created DESC"

Output:
    Results are displayed as a markdown table with key, status, priority, versions, and title.
`);
}

async function main() {
    const args = process.argv.slice(2);

    if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
        printHelp();
        process.exit(0);
    }

    const useRawJql = args[0] === '--jql';
    if (useRawJql) {
        args.shift();
    }

    if (args.length === 0) {
        console.error('Error: No search query provided.');
        printHelp();
        process.exit(1);
    }

    const query = args.join(' ');
    const token = getToken();
    const jql = buildJql(query, useRawJql);
    const response = await searchJira(jql, token);
    const output = formatAsMarkdownTable(response);
    console.log(output);
}

main().catch(err => {
    console.error(`Unexpected error: ${err.message}`);
    process.exit(1);
});
